### PR TITLE
e2e test of cops jobs

### PR DESCRIPTION
**Pull request by omehes**

### CODE REVIEW

**Comment by omehes** [comment]
The fail is not related to my current code. It is again to do with search in a book. When I re-run the same test, it PASSED. No issue here!

File: tests/cops_ui/test_cops_e2e.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+        # THEN we should get successful connection
+        assert response.status_code == 200
+
+
+def test_job_results(selenium, cops_base_url, cops_api_url):
+
+    sleep(2)
+
+    while True:
```
Comment (rnathuji) [inline tests/cops_ui/test_cops_e2e.py]: Will this spin forever potentially if any of the jobs fail and never reach `completed`, or does the test harness kill it after some point? If the former, perhaps we can give up after so many iterations (e.g. if we want to give up after X minutes, give up after X / sleep_time).

File: fixtures/webview.py [hunk 2]
```diff
@@ -469,3 +479,24 @@ def s3_books_titles():
...
+def cops_api_url(cops_base_url):
+    """Return URL for cops api"""
+    api_url = f"{cops_base_url}/api/jobs"
+    return api_url
+
+
+@pytest.fixture
+def create_new_pdf_job():
+    """  Return xpath of the 'Create a new pdf job button' """
+    create_new_pdf_job = '// *[ @ id = "app"] / div[1] / main / div / div / div / div / div[1] / button / span / span'
```
[the remark this answers is not in the document]
Comment (m1yag1) [reply fixtures/webview.py]: Additionally, fixtures are typically used to setup the test. They provide the "test harness" in which you pass in things the test will need. The name of this function makes me think it's a function meant to create a new pdf.

File: tests/cops_ui/test_cops_e2e.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        pytest.exit("<<<<<<<<<< HTTP Error: COPS URL IS INVALID OR UNREACHABLE >>>>>>>>>>")
+    except URLError:
+        pytest.exit("<<<<<<<<<< URL Error: COPS URL IS INVALID OR UNREACHABLE >>>>>>>>>>")
+    except ConnectionError:
+        pytest.exit("<<<<<<<<<< Connection Error: COPS URL IS INVALID OR UNREACHABLE >>>>>>>>>>")
+
+    else:
+
+        # opening 'CREATE A NEW PDF JOB' dialog
+        selenium.get(cops_base_url)
```
Comment (m1yag1) [inline tests/cops_ui/test_cops_e2e.py]: Is there a reason we aren't using a page object here?

File: fixtures/webview.py [hunk 4]
```diff
@@ -399,6 +399,16 @@ def rex_base_url(request):
         return base_url
 
 
+@pytest.fixture
```
Comment (m1yag1) [inline fixtures/webview.py]: Should this test live in cnx-automation?

File: tests/cops_ui/test_cops_e2e.py (hunk 3 above)
Comment (m1yag1) [reply tests/cops_ui/test_cops_e2e.py]: Remember the quote Otto,

> If you have WebDriver APIs in your test methods, You're Doing It Wrong. Simon Stewart

File: tests/cops_ui/test_cops_e2e.py (hunk 1 above)
Comment (omehes) [reply tests/cops_ui/test_cops_e2e.py]: Yes. You are right. I need to add some kind of timeout. Have to find a good solution.

File: tests/cops_ui/test_cops_e2e.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        # tabbing through fields and inputting colID, style and server
+        selenium.find_element_by_tag_name("body").send_keys(Keys.TAB, colid)
+        selenium.find_element_by_tag_name("body").send_keys(Keys.TAB)
+        selenium.find_element_by_tag_name("body").send_keys(Keys.TAB, style)
+        selenium.find_element_by_tag_name("body").send_keys(Keys.TAB, bserver)
+
+        # clicking 'CREATE' button to start the jobs
+        create_button_xp = selenium.find_element_by_xpath(create_button)
+        create_button_xp.click()
+        create_button_xp.click()
```
[the remark this answers is not in the document]
Comment (omehes) [reply tests/cops_ui/test_cops_e2e.py]: For some reason 1 click does not trigger the action. Not sure why. I didn't investigate it too much, I can give it a try.

File: tests/cops_ui/test_cops_e2e.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    while True:
+
+        api_page = urllib.request.urlopen(cops_api_url).read()
+
+        # loading cops json file and extracting required data
+        api_jdata = json.loads(api_page)
+
+        newest0 = api_jdata[0]
+        newest1 = api_jdata[1]
```
[the remark this answers is not in the document]
Comment (omehes) [reply tests/cops_ui/test_cops_e2e.py]: Yeah. I was experimenting with some if elif conditions. I should give it another try and check out your suggestions. Thanks

File: fixtures/webview.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (omehes) [reply fixtures/webview.py]: I couldn't find any ID there. And yes, I will rename the fixture. Now that you say it, it really looks like I am creating pdf with it...

File: tests/cops_ui/test_cops_e2e.py (hunk 3 above)
Comment (omehes) [reply tests/cops_ui/test_cops_e2e.py]: I am not page object but I can investigate.

File: fixtures/webview.py (hunk 4 above)
Comment (omehes) [reply fixtures/webview.py]: You mean it should be in its own repo? Like cops-automation? How would I do it?

**Comment by omehes** [comment]
@m1yag1 @rnathuji I made some changes as per your notes (not everything is fixed yet). If you have time, pls have a look and comment. 
checks failure is not related to my changes. If I rerun it, it passes. weird...

File: tests/cops_ui/test_cops_e2e.py (hunk 1 above)
Comment (rnathuji) [reply tests/cops_ui/test_cops_e2e.py]: @omehes - I took a look at the timeout mechanism in this commit and it looks like it should address the issue  :+1:

**Comment by omehes** [comment]
I created a new PR. closing this!

**Pull request closed without merging**